Add Ansible Repository refresh output page

Glutexo opened the pull request:
Create a custom display mode for Ansible Repository show page with last refresh output capture. If an Ansible Tower Repository has this capture from its last failed refresh, the Status row becomes a link to this subpage.

Presenting this information is a workaround for cloning repositories from HTTPS servers with untrusted SSL certificates. Tower API does not offer an option to disable SSL check, so with the stdout user can at least see, why the cloning actually failed.

Along with RSpec examples for this feature, refactor a bit other tests related to the page refresh button:

* Increase test coverage of the AJAX content replacement used
* Test Playbooks nested list subpage
* Rename examples with rather misleading titles

https://bugzilla.redhat.com/show_bug.cgi?id=1513616

Prior merging this PR, following PRs are required to be merged:

* [x] manageiq/manageiq#17290 model change
* [x] manageiq/manageiq-providers-ansible_tower#72 not necessarily required, but this PR doesn’t make sense without it

/cc @ZitaNemeckova, @hstastna for review

hstastna commented:
@Glutexo Could you please add some screenshots if there is some change in the UI? I'm not sure if I understand your change. Thanks!

hstastna commented:
And in specs, I would better use
```
is_expected.to have_http_status(200)
```
instead of `is_expected.to have_http_status 200`
but that's just a detail.

mzazrivec commented on app/controllers/ansible_repository_controller.rb:
```diff
@@ -97,6 +111,12 @@ def new
   end
 
+  def show_output
+    drop_breadcrumb(:name => "Refresh output", :url => show_output_link)
```
Missing gettext.

Glutexo commented:
I fixed the missing gettext spotted by @mzazrivec, squashed it and rebased onto current upstream master.

@hstastna As for the parentheses in the `have_http_status` method call: I agree with you, however I mimicked the style of the calls that is already there. So I leave it as it is, re-styling can be done in bulk if necessary.

Technically, the `expect(…).to` is a method call too, so it should be `expect(response).to(have_http_status(200))`…

hstastna commented on spec/controllers/ansible_repository_controller_spec.rb:
```diff
@@ -97,19 +121,82 @@
+      end
+
+      it 'calls show_output method' do
+        controller.params[:display] = 'output'
```
I see that you set `controller.params[:display]` in more `it` blocks exactly like this. You can use `context` blocks and set `params[:display]` in them by using `let` for lazy evaluation and `it` blocks then. For example like this:
```
context 'calling show_output method' do
  let(:display) { 'output' }

   it 'calls show_output method' do
     receives_init_show
     expect(controller).to receive(:show_output)
     renders
     controller.send(:button)
  end
end

``` 
If you do so, you may also want to change line 125 https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/files#diff-62ca34af6656484e37abce1fb2fe1461R125 like this, to set `params[:display]` for the spec properly:
```
      let(:params) { {:action => "button", :pressed => "ansible_repository_reload", :display => display} }
```

Glutexo commented:
When I created this PR, there was the same toolbar on subpages (custom display mode) as on main show page. Only there was an additional back arrow button. When I now want to test it after rebasing, I don’t see the toolbar buttons anymore. There is only the back arrow button.

Is it possible that toolbar behavior changed? I added a custom display mode in this PR and along with that a piece of code handling the refresh button. If the button is really no longer there, I can remove this newly introduced code and thus reduce complexity. Also in such case the [refresh button handling](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/ansible_repository_controller.rb#L43) could be simplified, not in this PR though.

Maybe @mzazrivec or @himdel could know?

himdel commented on app/controllers/ansible_repository_controller.rb:
```diff
@@ -40,11 +44,21 @@ def button
+
+      template = 'show'
+      if self.class.display_methods.include?(params[:display])
+        display_nested_list(params[:display])
```
how is `display_nested_list` relevant here? Are you introducing any nested lists?

Glutexo replied:
@himdel This is a piece taken from the `show` method in a not very DRY way.

The `show` method doesn’t [reach]((https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/mixins/generic_show_mixin.rb#L38)) the `custom_display_call` at all, so at least this one would have to remain here. Because of that and some other struggles I decided to leave only what’s necessary here from the `show` method.

That is: `init_show` and those two `if`s. The first is for `playbooks` subpage, which is a nested list, the second is for the new `output` subpage. I am however not sure whether it decreases complexity – the whole `show` method is not called here – or increases it – there are two decision points introduced.

However, since the Refresh button disappeared, this might not be actually needed at all. Moreover, the existing code there could be safely removed.

himdel commented on spec/controllers/ansible_repository_controller_spec.rb:
```diff
@@ -103,19 +127,93 @@
+        end
+      end
+
+      def receives_init_show
```
Cc @skateman , I think you may have some comments here..

skateman replied:
@Glutexo methods in a spec are evil, can you rewrite them using shared examples and shared contexts?

Glutexo replied:
@himdel @skateman No problem, I’ll rewrite it. Would you please provide me with a link to some resource explaining the evilness of methods in specs?

skateman replied:
TL;DR: mixing two patterns (rspec ruby vs pure ruby) is never a good idea
If you want a longer explanation, I can try next week if that's okay for you :wink:

himdel commented:
@Glutexo you may need to explicitly set `@center_toolbar = 'ansible_repository_center'`, so that the right toolbar gets picked up for the screen.

Glutexo commented:
I bisected the state before rebase and after that. The refresh button disappeared with the commit https://github.com/ManageIQ/manageiq-ui-classic/pull/3721/commits/3c7f204ef5e7e1a9dd3ab3f35a4038121f1740d1 by @hstastna. From my first glimpse at the PR that contains it (https://github.com/ManageIQ/manageiq-ui-classic/pull/3721/commits/3c7f204ef5e7e1a9dd3ab3f35a4038121f1740d1) it looks like the refresh button should not have been there in the first place. I’ll investigate it a bit more.

Glutexo commented:
I get it now. The playbooks subpage used the same toolbar as its parent repository page. Because of that the _Policy › Add Tags_ button added tags to the repository and not to the playbooks. That was addressed and fixed by #3721.

However at the same time the repository toolbar contained a refresh button. This button appeared on the playbooks subpage too. There is [a piece of code](https://github.com/ManageIQ/manageiq-ui-classic/blob/37a9337bbd4383328f1d722b9318ece80efbaaa3/app/controllers/ansible_repository_controller.rb#L43) that handles the button press. Thanks to [the magic in the generic show mixin](https://github.com/ManageIQ/manageiq-ui-classic/blob/37a9337bbd4383328f1d722b9318ece80efbaaa3/app/controllers/mixins/generic_show_mixin.rb#L3) it works for the playbooks subpage correctly too.

Still there is some rather obscure manipulation with the `@display` and `params[:display]` values: https://github.com/ManageIQ/manageiq-ui-classic/blob/37a9337bbd4383328f1d722b9318ece80efbaaa3/app/controllers/ansible_repository_controller.rb#L44 followed directly by https://github.com/ManageIQ/manageiq-ui-classic/blob/37a9337bbd4383328f1d722b9318ece80efbaaa3/app/controllers/mixins/generic_show_mixin.rb#L121. This still [doesn’t work](https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/files#diff-9b524579bdd0b54c65c5c3c0c778d2efR47) too great.

For a mix of reasons, this piece of code didn’t work automatically for [the new stdout subpage](https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/files#diff-9b524579bdd0b54c65c5c3c0c778d2efR115) I introduced. Because of that I made [some controversial modifications](https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/files#r194043356) to [the refresh button handling](https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/files#diff-9b524579bdd0b54c65c5c3c0c778d2efR46). But since the button is no longer there, although not necessary intentionally, this piece of code is not necessary and can be removed from this PR.

Glutexo commented:
So, the refresh button [is gone](https://github.com/ManageIQ/manageiq-ui-classic/pull/3721) for subpages (be in intentional or not). Thanks to that I could simplify this PR. Removed changes to the refresh button handling making it work for the new subpage as well as the related tests. Those can be added in case the refresh button [would return](https://github.com/ManageIQ/manageiq-ui-classic/pull/3721#issuecomment-396900036).

Still this is not mergeable, because of dependencies on other PRs. It is not however at least simple enough to be reviewed.

cc @himdel @hstastna

hstastna commented:
@Glutexo It looks good, I just cannot access https://user-images.githubusercontent.com/1161096/39128873-3ca77670-4709-11e8-843a-bf301fa35b22.png How can I do this? It's not clear to me. Thank you! :)

hstastna commented:
@Glutexo Travis is failing here and it looks like it is related to your changes. Specs are failing for me locally, too, with your changes (they pass without your changes). Could you check it, please? Thank you! :+1:

Glutexo commented:
Okay, I’ll chech more deeply.

Glutexo commented:
# Failing specs #

Oh, of course the specs are failing! It’s because these changes depend on a model/schema update: Please checkout https://github.com/Glutexo/manageiq/tree/add_configuration_script_source_last_update_stdout and https://github.com/Glutexo/manageiq-schema/tree/add_configuration_script_source_last_update_stdout to update the database schema and the Rails model.

Unfortunately I am currently not able to quickly test it locally, because Bundler and updates fail for me for some unknown reason. But that above is necessary to test both by rspec and manually.

# How to test the UI behavior #

## Prerequisities ##

* You have to have some Ansible Repositories in your database. I assume you already have those.
* You have to have your database schema updated, so you have the `last_update_stdout` column in the `configuration_script_source` table.

## Steps ##

* Fill in something into the `last_update_stdout` column of some Ansible repository.
* Find this repository in the list (Automate › Ansible › Repositories) and click it to open its show page.
* See the _Status_ row containing _success_/_failed_ and check it is a link.
* Click the row to open a separate page with the output.
* Check that there is the column contents displayed.
* Set the column value to `NULL`.
* Refresh the page.
* Check that no error (500) is raised and there is a friendly notice on the page.
* Go back to the main repository show page and refresh it.
* See the _Status_ row again and check it is not a link.

hstastna commented:
Of course I've updated the db schema. And specs are still failing.

Glutexo commented:
@hstastna I know you said you updated the schema, but still I’d like to verify you don’t have the same problem as I did.

I double- (triple-) checked and got failing specs too. This was really because of the missing column, at least on my machine. After a few moments of despair I realized that I have to overload the `manageiq-schema` gem in my custom `bundler.d` Gemfile to pick my branch with the additional migration. When I fixed this and ran `rake test:vmdb:setup` the tests magically turned green.

hstastna commented:
@Glutexo Thank you very much for your detailed explanation. Yes, sometimes we can forget about some specific things which we should do..  But.. I did the same and it didn't help :sweat_smile: Maybe I'm still doing something wrong. Anyway, @skateman is better expert for specs ;)

skateman commented:
Try from the manageiq core repo:
```bash
RAILS_ENV=test bin/rake db:migrate
```
Might help, but just a guess :wink:

Glutexo commented:
@hstastna Just to be sure. When you run the `ansible_repository_controller.spec` and it fails, takes look into `test.log` please. What does that say? Just to be sure it‘s not _method_missing_ like this:

```
[----] F, [2018-06-20T14:49:47.574607 #50762:3fd99e43f7bc] FATAL -- : Error caught: [ActionView::Template::Error] undefined method `last_update_stdout' for #<ManageIQ::Providers::EmbeddedAnsible::AutomationManager::ConfigurationScriptSource:0x00007fb347c30418>
Did you mean?  last_updated_on
```

which would mean that the schema is bad. If it’s something different, then I am wrong with all those schema-rants for the whole time.

Glutexo commented:
@Fryguy [requested](https://github.com/ManageIQ/manageiq/pull/17290#issuecomment-398781641) a different name for the new column. I changed it in the [schema](https://github.com/ManageIQ/manageiq-schema/pull/187) and in the [model](https://github.com/ManageIQ/manageiq/pull/17290#issuecomment-398781641) as well as [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/3762/commits/c847860062867c4618e323b781a350c99f33c988).

Glutexo commented:
@hstastna I’d prefer not to split the commit to have executive code in a different one than tests. I like that the commit is atomic and tested by itself. Moreover, the spec changes are so little that it doesn't make much sense to extract those. Sorry. 😔

May I ask you to check running the specs with the new column name?

tumido commented:
All related PRs merged now.

@miq-bot remove_label pending core

himdel commented:
Tested, looks good :+1: :)

The only thing:

> Fill in something into the last_update_stdout column of some Ansible repository.

this is called `last_update_error` now.
